bigquery_etl/cli/query.py

```diff
@@ -1043,6 +1051,37 @@ def create_query_session(
     return job.session_info.session_id
 
 
+def extract_and_run_temp_udfs(query_text: str, project_id: str, session_id: str) -> str:
+    """Create temp udfs in the session and return the query without udf definitions.
+
+    Does not support dry run because the query will fail dry run if udfs aren't defined.
+    """
+    udfs = get_local_routines_from_text(query_text)
+
+    if len(udfs) == 0:
+        return query_text
+
+    updated_query = query_text
+
+    for udf in udfs:
+        udf_name = get_local_routine_name(udf)
+        updated_query = updated_query.replace(
+            udf, f"-- temp udf created in session: {udf_name}\n"
+        )
+
+    udf_def_statement = "\n".join([udf for udf in udfs])
+
+    client = bigquery.Client(project=project_id)
+
+    job_config = bigquery.QueryJobConfig(
+        use_legacy_sql=False,
+        connection_properties=[bigquery.ConnectionProperty("session_id", session_id)],
+    )
+    client.query_and_wait(udf_def_statement, job_config=job_config)
+
+    return updated_query
```

Comment:
When a session is involved, would it perhaps be simpler to just split non-script queries into separate statements (if any), and run each of those statements separately, with only the final statement query having the destination table set?

Reply:
Yes that's a good point.  I was initially accounting for non-udf statements that might be in the query especially since it can't be dry run but I'm not sure there are any other possible statement types

Reply:
I ran sqlparse on the output of `[f for f in paths_matching_name_pattern(None, "sql", None) if f.name != "script.sql"]` and confirmed the only other statements are udfs so it's safe to take out the temp udf specific code.  I'm thinking of still running those as one query though since it's easier to look through in the job logs.